tensorflow/lite/micro/kernels/squeeze_test.cc

```diff
@@ -0,0 +1,193 @@
+  constexpr int tensors_size = inputs_size + outputs_size;
+  T* initial_output_data = reinterpret_cast<T*>(output_data_raw);
+  TfLiteTensor tensors[tensors_size] = {
+      CreateTensor<T, tensor_input_type>(input_data, input_dims,
+                                         "input_tensor"),
+      CreateTensor<T, tensor_input_type>(initial_output_data,
+                                         initial_output_dims, "output_tensor"),
+  };
+  TfLiteContext context;
+  PopulateContext(tensors, tensors_size, &context);
+  ::tflite::ops::micro::AllOpsResolver resolver;
+  const TfLiteRegistration* registration =
+      resolver.FindOp(tflite::BuiltinOperator_SQUEEZE, 1);
+  TF_LITE_MICRO_EXPECT_NE(nullptr, registration);
+
+  int inputs_array_data[] = {1, 0};
+  TfLiteIntArray* inputs_array = IntArrayFromInts(inputs_array_data);
+  int outputs_array_data[] = {1, 1};
+  TfLiteIntArray* outputs_array = IntArrayFromInts(outputs_array_data);
+
+  // TODO(See TfLiteSqueezeParams): We can't have dynamic data, at least not
+  // yet. Thus we allocate space for outputs_dims_array in the temporaries
+  // buffer.
+  int outputs_dims_array_data[max_num_dims + 1] = {
+      max_num_dims, 0, 0, 0, 0, 0, 0, 0, 0};
+  TfLiteIntArray* outputs_dims_array =
+      IntArrayFromInts(outputs_dims_array_data);
+
+  TfLiteSqueezeParams params;
+  TF_LITE_MICRO_EXPECT_LE(squeeze_dims.size(), max_num_dims);
+  params.num_squeeze_dims = squeeze_dims.size();
+  for (int i = 0; i < params.num_squeeze_dims; ++i) {
+    params.squeeze_dims[i] = squeeze_dims.begin()[i];
+  }
+
+  // Run op
+  TfLiteNode node;
+  node.inputs = inputs_array;
+  node.outputs = outputs_array;
+  node.temporaries = outputs_dims_array;
+  node.user_data = nullptr;
+  node.builtin_data = &params;
+  node.custom_initial_data = nullptr;
+  node.custom_initial_data_size = 0;
+  node.delegate = nullptr;
+  TF_LITE_MICRO_EXPECT_NE(nullptr, registration->invoke);
+  TF_LITE_MICRO_EXPECT_EQ(kTfLiteOk, registration->invoke(&context, &node));
+
+  // Check the squeezed dimensions are as expected
+  constexpr int outputs_index = 1;
+  TfLiteIntArray* output_dims = tensors[outputs_index].dims;
+
+  TF_LITE_MICRO_EXPECT_EQ(expected_output_dims->size, output_dims->size);
+  for (int i = 0; i < output_dims->size; ++i) {
+    TF_LITE_MICRO_EXPECT_EQ(expected_output_dims->data[i],
+                            output_dims->data[i]);
+  }
+
+  // Check that squeeze does not mutate the data
+  const T* output_data = GetTensorData<T>(&tensors[outputs_index]);
+  int flat_size = 1;
+  for (int i = 0; i < output_dims->size; ++i) {
+    flat_size *= output_dims->data[i];
+  }
+  for (int i = 0; i < flat_size; ++i) {
+    TF_LITE_MICRO_EXPECT_NEAR(expected_output.begin()[i], output_data[i],
+                              1e-5f);
+  }
+}
+
+}  // namespace
+}  // namespace testing
+}  // namespace tflite
+
+TF_LITE_MICRO_TESTS_BEGIN
+
+#define TEST_SQUEEZE(...)                                           \
```

Comment:
Does testing each type gain you anything here?  I would prefer to avoid macros and instead just test for float for this op.  For ops which perform differently for various input types, we find it's clearer to have a test for each input type, since test failures can more easily be traced back to the specific case that fails.

Reply:
Done. Yes testing types other than float isn't necessary for squeeze op. Thanks for the info about the macro.